### ISSUE TITLE
Missing distribution via Maven Central Repository

### ISSUE THREAD

**Issue by PAX523** [capstone-engine/capstone#605]
As a Java developer I prefer external project library dependencies that can be obtained as Maven dependencies: http://search.maven.org/

This means, it would be very nice if you provide your Java binding (inclusive native library imbedded) as an artifact on Maven Central. Is this possible for you in future?

**Comment by aquynh** [comment]
Sure, can you contribute this part?

**Comment by PAX523** [comment]
Alright! I'd be glad to! Please allow me some time.

**Comment by PAX523** [comment]
It seems that the Java source have to be converted into Maven projects. Among other things this means that I would have to move the sources directory into another structure. Is this okay for you?

Furthermore, it's common practice to use a different package naming convention. Usally you use the website's URL for this purpose. I suggest:

    org.capstoneengine

If you agree with these measures then I'm going to fork your project in the next these and finally ask for a pull request?

**Comment by aquynh** [comment]
backward compatibility is important. does this change require any modification to existing Java apps using current Java binding?

**Comment by PAX523** [comment]
Yes! The Java binding would become incompatible to existing code! Except we don't touch the package structure. But even in this case we have to make sure that all current Java pull request are merged into master, completely, and no further project forks are being done until I converted the project into Maven structure.

**Comment by PAX523** [comment]
I would separate into 3 artifacts:

- Java binding
- Native libs for 32 bit
- Native libs for 64 bit

I will add JNA library as Maven dependency into the Maven-POM file. Users who use maven need to add the Java binding artifact and the desired native library artifact to their dependencies. Then they get everything they need to use capstone.

**Comment by aquynh** [comment]
@danghvu, do you have any comments?

**Comment by PAX523** [comment]
I investigated a little bit. We don't need two native artifacts. One with both system architectures is enough. The libs must be in following folder structure (for windows libs):

- ``win32-x86/capstone.dll``
- ``win32-x86-64/capstone.dll``

I couldn't find the natives for Linux and Mac. But at first we could provide one natives artifact for Windows, only. This means we would have 2 artifacts:

- Java binding
- Native libs (32/64) for Windows

**Comment by PAX523** [comment]
There's a problem: Native libraries cannot be packaged into JARs during runtime. They need to be extracted, first. See: https://www.adamheinrich.com/blog/2012/12/how-to-load-native-jni-library-from-jar/

**Comment by aquynh** [comment]
if there is no choice but breaking compatibility, then we should keep it minimal.

can you elaborate (with example) on how your change will break compatibility of existing Java apps?

**Comment by aquynh** [comment]
so it seems the only change to existing apps is to modify all the calls to "import"?

**Comment by PAX523** [comment]
Correct! If we decide to change the package structure (but we don't necessarily need to) then existing apps would have compile errors and the maintainers have to modifiy the imports.

**Comment by aquynh** [comment]
so i understand that we dont need to change the structure, but supporting Maven is still possible? then we dont break anythings, so please go ahead and do what you want.

**Comment by PAX523** [comment]
Before I can go on we have to make sure that there are no pending pull requests that modify the Java binding. Because Maven projects require a special folder structure. I have to move the sources (Don't worry! Package structure still keeps the same!). If somebody does a pull request then you need to merge all changes manually. I don't believe that the merge tools realize how the sources were moved.

Before I start all current pull request must be merged into master, first. Furthermore, no new forks must be done before the project is converted into Maven structure.

I found this pending pull request which concerns Java code: #535

We run into troubles if we try to merge this into master.

**Comment by aquynh** [comment]
it seems the author gave up on that PR #535 long ago. so if possible, please also pick up that abandoned work for your PR. thanks.

**Comment by PAX523** [comment]
The problem is that I don't really understand whether the changes in #535 are alright or not. I don't understand what the changes do.

**Comment by aquynh** [comment]
oh that is because your PR is for the master branch, and you dont have that API there.

the "next" branch is where all the latest development happens, which adds a lot of more advanced features, and will be released as v4.0 soon. the #535 is to support new API only available in "next". see here for more info: http://www.capstone-engine.org/op_access.html

i will find someone to help to review your Maven PR, thanks!

**Comment by PAX523** [comment]
Does it mean that I also have to port the Maven stuff into the "next" branch? Or will you do that yourself?

Please do me a favor: Please try to execute Makefile of the Java binding on your Linux machine and tell me whether it fails or not. I'm not sure if my changes were enough.

**Comment by PAX523** [comment]
I prepared everything as far as possible. For the Maven artifact release management you need to investigate a little bit. Some steps are only possible from your account.

You need an account on sonatype: https://issues.sonatype.org/secure/Dashboard.jspa

You have to install the PGP program on your local machine: http://www.sonatype.org/nexus/2015/01/08/deploy-to-maven-central-repository/

The setup and the handling with Maven is described here: http://central.sonatype.org/pages/apache-maven.html

If you need further help or have any questions feel free to ask me.

**Comment by kabeor** [comment]
Close this issue for now because of legacy, we are preparing to release Capstone 5.0, plz feel free to open a new issue if that still has this issue.
thx :)
